### PR TITLE
Add a new command doctor to check if some wrong configurations on gitea instance

### PR DESCRIPTION
**Pull request by lunny**
When gitea's path is not changed.

图片

When gitea's path changed.
图片

### CODE REVIEW

**Comment by guillep2k** [comment]
This is about #8781

**Review by 6543 — APPROVED** [review]

**Comment by jolheiser** [comment]
@lunny Is this considered WIP currently, or implement the one check and add more with other PRs?

File: cmd/doctor.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+		firstline = scanner.Text()
+		if !strings.HasPrefix(firstline, "#") {
+			break
+		}
+	}
+
+	// command="/Volumes/data/Projects/gitea/gitea/gitea --config
+	if len(firstline) > 0 {
```
Comment (guillep2k) [inline cmd/doctor.go:100]: You could just use a regexp for this:
```
exp := regexp.MustCompile(`^[ \t]*(?:command=")([^ ]+) --config='([^']+)' serv key-([^"]+)",(?:[^ ]+) ssh-rsa ([^ ]+) ([^ ]+)[ \t]*$`)

// command="/home/user/gitea --config='/home/user/etc/app.ini' serv key-999",option-1,option-2,option-n ssh-rsa public-key-value key-name 
res := exp.FindAllStringSubmatch(firstline, -1)

exePath := res[1]  // => /home/user/gitea
iniPath := res[2]  // => /home/user/etc/app.ini
keyID := res[3]    // => 999
key := res[4]      // => public-key-value
keyName := res[5]  // => key-name
```
So you can check the `exe` _and_ the `ini` paths, and even rewrite them if you want.

https://regex101.com/r/JZAMwU/3

**Comment by lunny** [comment]
@jolheiser Not WIP, I just want to add a framework and add more checkes on late PRs.

File: cmd/doctor.go (hunk 1 above)
Comment (lunny) [reply cmd/doctor.go:100]: merged some necessary codes.

File: cmd/doctor.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+		exp := regexp.MustCompile(`^[ \t]*(?:command=")([^ ]+) --config='([^']+)' serv key-([^"]+)",(?:[^ ]+) ssh-rsa ([^ ]+) ([^ ]+)[ \t]*$`)
+
+		// command="/home/user/gitea --config='/home/user/etc/app.ini' serv key-999",option-1,option-2,option-n ssh-rsa public-key-value key-name
+		res := exp.FindAllStringSubmatch(firstline, -1)
```
Comment (guillep2k) [inline cmd/doctor.go]:
```suggestion
		res := exp.FindStringSubmatch(firstline)
		if res == nil {
			fmt.Println("...some warning....")
			// check next line?
		}
```

`res` must be checked for `nil`. If the line doesn't match, `nil` is returned. In my case I have some gitea settings mixed with some personal settings (`ssh-rsa ...`) and this causes a panic. I'd just print a warning if a line does not match Gitea's command.

EDIT: I see that you're not checking all lines. It's not a problem, but checking beyond the first line could be useful.

EDIT2: Sorry, it's `FindStringSubmatch()` and not `FindAllStringSubmatch()` as I've mistakenly said in a comment earlier. This means sub-indexing `giteaPath` and `iniPath` is not required.

**Comment by 6543** [comment]
because it `add a framework` we need a litle bit of docu to be usefull

**Review by 6543 — CHANGES_REQUESTED** [review]
some docu hints

File: cmd/doctor.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+func exePath() (string, error) {
+	file, err := exec.LookPath(os.Args[0])
+	if err != nil {
+		return "", err
+	}
+	return filepath.Abs(file)
+}
+
+func runDoctorLocationMoved(ctx *cli.Context) ([]string, error) {
+	if setting.SSH.StartBuiltinServer {
```
Comment (sapk) [inline cmd/doctor.go]: This should also be disabled if SSH_CREATE_AUTHORIZED_KEYS_FILE=false 
```suggestion
	if setting.SSH.StartBuiltinServer || !setting.SSH.CreateAuthorizedKeysFile {
```
Similar like to the test here: https://github.com/go-gitea/gitea/blob/dce22efbee6a04b8c19348dac831cd88b18ac07c/models/ssh_key.go#L653

**Review by sapk — CHANGES_REQUESTED** [review]

**Comment by 6543** [comment]
@lunny can you add docs about this feature?

**Comment by guillep2k** [comment]
> 
> 
> @lunny can you add docs about this feature?

I'd suggest not delay this anymore, so others could start adding functions to it as well. 😉

Docs can wait to another PR, IMHO.

**Review by 6543 — APPROVED** [review]

**Comment by lunny** [comment]
@sapk @guillep2k done.

**Comment by 6543** [comment]
@lunny thanks

**Review by guillep2k — APPROVED** [review]

**Review by sapk — APPROVED** [review]

**Pull request merged**